libs/ml/benchmark/embeddings/embeddings.cpp

```diff
@@ -0,0 +1,204 @@
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "logging/logging.hpp"
+#include "math/tensor.hpp"
+#include "ml/core/graph.hpp"
+#include "ml/layers/fully_connected.hpp"
+#include "ml/ops/activations/relu.hpp"
+#include "ml/ops/loss_functions/mean_square_error_loss.hpp"
+#include "ml/optimisation/adam_optimiser.hpp"
+#include "ml/optimisation/lazy_adam_optimiser.hpp"
+
+#include "benchmark/benchmark.h"
+
+#include <memory>
+#include <string>
+
+//////////////////////////
+/// reusable functions ///
+//////////////////////////
+
+template <typename TypeParam>
+std::shared_ptr<fetch::ml::Graph<TypeParam>> PrepareTestGraph(
+    typename TypeParam::SizeType embedding_dimensions, typename TypeParam::SizeType n_datapoints,
+    std::string &input_name, std::string &label_name, std::string &error_name)
+{
+  std::shared_ptr<fetch::ml::Graph<TypeParam>> g(std::make_shared<fetch::ml::Graph<TypeParam>>());
+
+  input_name = g->template AddNode<fetch::ml::ops::PlaceHolder<TypeParam>>("", {});
+
+  std::string output_name = g->template AddNode<fetch::ml::ops::Embeddings<TypeParam>>(
+      "Embeddings", {input_name}, embedding_dimensions, n_datapoints);
+
+  label_name = g->template AddNode<fetch::ml::ops::PlaceHolder<TypeParam>>("", {});
+  error_name = g->template AddNode<fetch::ml::ops::MeanSquareErrorLoss<TypeParam>>(
+      "Error", {output_name, label_name});
+
+  return g;
+}
+
+template <typename T, fetch::math::SizeType B, fetch::math::SizeType S, fetch::math::SizeType D,
+          fetch::math::SizeType E>
+void BM_Setup_And_Train_Embeddings_Adam(benchmark::State &state)
+{
+  using SizeType   = fetch::math::SizeType;
+  using DataType   = T;
+  using TensorType = fetch::math::Tensor<DataType>;
+
+  fetch::SetGlobalLogLevel(fetch::LogLevel::ERROR);
+
+  SizeType batch_size           = B;
+  SizeType embedding_dimensions = S;
+  SizeType n_datapoints         = D;
+  SizeType n_epochs             = E;
+
+  auto learning_rate = DataType{0.1f};
+
+  // Prepare data and labels
+  TensorType data({1, batch_size});
+  TensorType gt({embedding_dimensions, 1, batch_size});
+  data.FillUniformRandomIntegers(0, static_cast<int64_t>(n_datapoints));
+  gt.FillUniformRandom();
+
+  for (auto _ : state)
+  {
+    // make a graph
+    std::string input_name;
+    std::string label_name;
+    std::string error_name;
+    auto        g = PrepareTestGraph<TensorType>(embedding_dimensions, n_datapoints, input_name,
```

Comment:
you seem to be including graph construction inside the benchmarking loop. Do you actually want to be benchmarking this?